pkg/reconciler/cluster/workloadclusterheartbeat/cluster_manager.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workloadclusterheartbeat
+
+import (
+	"context"
+	"time"
+
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/reconciler/cluster"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/third_party/conditions/util/conditions"
+)
+
+var _ cluster.ClusterReconcileImpl = (*clusterManager)(nil)
+
+type clusterManager struct {
+	heartbeatThreshold  time.Duration
+	enqueueClusterAfter func(*workloadv1alpha1.WorkloadCluster, time.Duration)
+}
+
+func (c *clusterManager) Reconcile(ctx context.Context, cluster *workloadv1alpha1.WorkloadCluster) error {
+	latestHeartbeat := time.Time{}
+	for _, c := range cluster.Status.Conditions {
+		if c.LastHeartbeatTime.Time.After(latestHeartbeat) {
+			latestHeartbeat = c.LastHeartbeatTime.Time
+		}
+	}
+	if latestHeartbeat.IsZero() {
+		// Never seen a heartbeat; for now, this means the syncer hasn't heartbeated -- in the future, this should be un-Ready.
+		return nil
+	}
+
+	if time.Since(latestHeartbeat) > c.heartbeatThreshold {
```

Comment:
(No action required) Maybe add a TODO to use something like a sliding window to avoid setting unready for a transient issue?

Reply:
My thinking was that the heartbeat interval would be some fraction of this controller's heartbeat threshold, say 10s. So the controller would only mark the cluster as not-`Ready` if N (=6) heartbeats were missed. Both of these would be configurable from either end, so you could tweak that to your liking based on the relative amount of transient issues you expect from clusters in the system.